File: backend/app/services/google_login_oauth.py
"""Google OpenID Connect flow for ordinary member sign-in.

This integration is intentionally separate from ``gmail_oauth``: it requests
only identity scopes, stores no Google access/refresh token, and converts a
successful callback into a short-lived, single-use local exchange code.
"""
from __future__ import annotations

import base64
from dataclasses import dataclass
import hashlib
import json
from pathlib import Path
import secrets
from threading import Lock
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

import jwt

from ..core.config import get_settings


_AUTH_URI = "https://accounts.google.com/o/oauth2/v2/auth"
_TOKEN_URI = "https://oauth2.googleapis.com/token"
_JWKS_URI = "https://www.googleapis.com/oauth2/v3/certs"
_SCOPES = "openid email profile"
STATE_COOKIE = "thaiperform_google_login_state"


class GoogleLoginOAuthError(RuntimeError):
    """Safe, non-secret error raised during member Google sign-in."""


@dataclass(frozen=True)
class OAuthClient:
    client_id: str
    client_secret: str
    auth_uri: str
    token_uri: str
    redirect_uri: str


@dataclass(frozen=True)
class GoogleIdentity:
    subject_id: str
    email: str
    display_name: str
    avatar_url: str


_pending: dict[str, tuple[float, str, str]] = {}
_pending_lock = Lock()
_login_codes: dict[str, tuple[float, int, str]] = {}
_login_codes_lock = Lock()


def _safe_next_path(value: str | None) -> str:
    candidate = (value or "").strip()
    if candidate.startswith("/") and not candidate.startswith("//"):
        return candidate
    return "/recommend"


def _read_json(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError, TypeError) as exc:
        raise GoogleLoginOAuthError("Google login credential file is unreadable") from exc
    if not isinstance(payload, dict):
        raise GoogleLoginOAuthError("Google login credential file is invalid")
    return payload


def load_client() -> OAuthClient:
    settings = get_settings()
    client_id = str(settings.google_login_client_id or "").strip()
    client_secret = str(settings.google_login_client_secret or "").strip()
    auth_uri = _AUTH_URI
    token_uri = _TOKEN_URI
    client_file = settings.google_login_client_file
    if (not client_id or not client_secret) and client_file:
        web = _read_json(Path(client_file)).get("web")
        if not isinstance(web, dict):
            raise GoogleLoginOAuthError("OAuth credential must be a Web application client")
        client_id = str(web.get("client_id") or "").strip()
        client_secret = str(web.get("client_secret") or "").strip()
        auth_uri = str(web.get("auth_uri") or auth_uri).strip()
        token_uri = str(web.get("token_uri") or token_uri).strip()
    if not client_id or not client_secret:
        raise GoogleLoginOAuthError("Google member login is not configured")
    return OAuthClient(
        client_id=client_id,
        client_secret=client_secret,
        auth_uri=auth_uri,
        token_uri=token_uri,
        redirect_uri=str(settings.google_login_redirect_uri).strip(),
    )


def client_configured() -> bool:
    try:
        load_client()
        return True
    except GoogleLoginOAuthError:
        return False


def start_authorization(next_path: str | None = None) -> tuple[str, str]:
    """Return the Google authorization URL plus state bound to the browser."""
    client = load_client()
    state = secrets.token_urlsafe(32)
    verifier = secrets.token_urlsafe(64)
    challenge = base64.urlsafe_b64encode(
        hashlib.sha256(verifier.encode("ascii")).digest()
    ).rstrip(b"=").decode("ascii")
    now = time.time()
    ttl = max(60, int(get_settings().google_login_state_ttl_seconds))
    with _pending_lock:
        for key, (created, _, _) in list(_pending.items()):
            if now - created > ttl:
                _pending.pop(key, None)
        _pending[state] = (now, verifier, _safe_next_path(next_path))
    query = urlencode(
        {
            "client_id": client.client_id,
            "redirect_uri": client.redirect_uri,
            "response_type": "code",
            "scope": _SCOPES,
            "state": state,
            "code_challenge": challenge,
            "code_challenge_method": "S256",
            "prompt": "select_account",
        }
    )
    return f"{client.auth_uri}?{query}", state


def complete_authorization(code: str, state: str) -> tuple[GoogleIdentity, str]:
    with _pending_lock:
        pending = _pending.pop(state, None)
    ttl = max(60, int(get_settings().google_login_state_ttl_seconds))
    if pending is None or time.time() - pending[0] > ttl:
        raise GoogleLoginOAuthError("Google login state is invalid or expired")
    client = load_client()
    token_payload = _post_form(
        client.token_uri,
        {
            "code": code,
            "client_id": client.client_id,
            "client_secret": client.client_secret,
            "redirect_uri": client.redirect_uri,
            "grant_type": "authorization_code",
            "code_verifier": pending[1],
        },
    )
    id_token = str(token_payload.get("id_token") or "").strip()
    if not id_token:
        raise GoogleLoginOAuthError("Google did not return an identity token")
    return _verify_id_token(id_token, client.client_id), pending[2]


def _verify_id_token(id_token: str, client_id: str) -> GoogleIdentity:
    try:
        signing_key = jwt.PyJWKClient(_JWKS_URI).get_signing_key_from_jwt(id_token)
        claims = jwt.decode(
            id_token,
            signing_key.key,
            algorithms=["RS256"],
            audience=client_id,
        )
    except Exception as exc:
        raise GoogleLoginOAuthError("Google identity token could not be verified") from exc
    if claims.get("iss") not in {"accounts.google.com", "https://accounts.google.com"}:
        raise GoogleLoginOAuthError("Google identity token has an invalid issuer")
    if claims.get("email_verified") is not True:
        raise GoogleLoginOAuthError("Google account email is not verified")
    subject = str(claims.get("sub") or "").strip()
    email = str(claims.get("email") or "").strip().lower()
    if not subject or not email:
        raise GoogleLoginOAuthError("Google identity is missing required fields")
    return GoogleIdentity(
        subject_id=subject,
        email=email,
        display_name=str(claims.get("name") or "").strip(),
        avatar_url=str(claims.get("picture") or "").strip(),
    )


def issue_login_code(user_id: int, next_path: str) -> str:
    raw = secrets.token_urlsafe(32)
    digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
    now = time.time()
    ttl = max(30, int(get_settings().google_login_code_ttl_seconds))
    with _login_codes_lock:
        for key, (created, _, _) in list(_login_codes.items()):
            if now - created > ttl:
                _login_codes.pop(key, None)
        _login_codes[digest] = (now, int(user_id), _safe_next_path(next_path))
    return raw


def consume_login_code(raw: str) -> tuple[int, str] | None:
    digest = hashlib.sha256((raw or "").encode("utf-8")).hexdigest()
    with _login_codes_lock:
        value = _login_codes.pop(digest, None)
    ttl = max(30, int(get_settings().google_login_code_ttl_seconds))
    if value is None or time.time() - value[0] > ttl:
        return None
    return value[1], value[2]


def _post_form(url: str, fields: dict[str, str]) -> dict[str, Any]:
    request = Request(
        url,
        data=urlencode(fields).encode("utf-8"),
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        method="POST",
    )
    try:
        with urlopen(request, timeout=20) as response:
            payload = json.loads(response.read().decode("utf-8"))
    except HTTPError as exc:
        raise GoogleLoginOAuthError(f"Google token exchange failed ({exc.code})") from exc
    except (URLError, OSError, ValueError) as exc:
        raise GoogleLoginOAuthError("Google token exchange failed") from exc
    if not isinstance(payload, dict):
        raise GoogleLoginOAuthError("Google token response is invalid")
    return payload

File: frontend/lib/server/google-login-oauth.ts
// Member Google OpenID Connect (separate from the admin Gmail sender client).
//
// Requests only identity scopes, never stores a Google access/refresh token,
// and converts a successful callback into a normal Application Backend server
// session. The OAuth exchange and ID-token verification happen server-side in
// Next.js route handlers; the browser only ever navigates.
import { createHash, randomBytes } from "node:crypto";
import { readFileSync } from "node:fs";
import { createRemoteJWKSet, jwtVerify } from "jose";
import { postForm } from "@/lib/server/oauth-http";

export const LOGIN_STATE_COOKIE = "thaiperform_google_login_state";
const SCOPES = "openid email profile";
const DEFAULT_AUTH_URI = "https://accounts.google.com/o/oauth2/v2/auth";
const DEFAULT_TOKEN_URI = "https://oauth2.googleapis.com/token";
const DEFAULT_JWKS_URI = "https://www.googleapis.com/oauth2/v3/certs";
const GOOGLE_ISSUERS = new Set(["accounts.google.com", "https://accounts.google.com"]);

export class GoogleLoginOAuthError extends Error {}

export interface GoogleLoginClient {
  clientId: string;
  clientSecret: string;
  authUri: string;
  tokenUri: string;
  jwksUri: string;
  redirectUri: string;
}

export interface GoogleIdentity {
  subjectId: string;
  email: string;
  displayName: string;
  avatarUrl: string;
}

export interface PendingLoginFlow {
  state: string;
  verifier: string;
  nextPath: string;
}

function envValue(name: string): string {
  return (process.env[name] ?? "").trim();
}

// fallow-ignore-next-line complexity -- Unreadable and malformed credential files must each fail closed.
function readClientFile(path: string | undefined): Record<string, unknown> | null {
  if (!path) return null;
  try {
    const parsed = JSON.parse(readFileSync(path, "utf8")) as Record<string, unknown>;
    return parsed && typeof parsed === "object" ? parsed : null;
  } catch {
    return null;
  }
}

function clientWeb(path: string | undefined): Record<string, unknown> | null {
  const parsed = readClientFile(path);
  if (!parsed) return null;
  const web = parsed.web as Record<string, unknown> | undefined;
  if (!web || typeof web !== "object") {
    throw new GoogleLoginOAuthError("OAuth credential must be a Web application client");
  }
  return web;
}

// fallow-ignore-next-line complexity -- Env-var and client-file fallbacks must resolve in one deterministic precedence order.
export function loadLoginClient(): GoogleLoginClient {
  const web = clientWeb(envValue("GOOGLE_LOGIN_CLIENT_FILE"));
  const clientId = envValue("GOOGLE_LOGIN_CLIENT_ID") || String(web?.client_id ?? "").trim();
  const clientSecret = envValue("GOOGLE_LOGIN_CLIENT_SECRET") || String(web?.client_secret ?? "").trim();
  const authUri = envValue("GOOGLE_LOGIN_AUTH_URI") || String(web?.auth_uri ?? "").trim() || DEFAULT_AUTH_URI;
  const tokenUri = envValue("GOOGLE_LOGIN_TOKEN_URI") || String(web?.token_uri ?? "").trim() || DEFAULT_TOKEN_URI;
  const jwksUri = envValue("GOOGLE_LOGIN_JWKS_URI") || DEFAULT_JWKS_URI;
  const redirectUri = envValue("GOOGLE_LOGIN_REDIRECT_URI");
  if (!clientId || !clientSecret || !redirectUri) {
    throw new GoogleLoginOAuthError("Google member login is not configured");
  }
  return { clientId, clientSecret, authUri, tokenUri, jwksUri, redirectUri };
}

export function safeNextPath(value: string | null | undefined): string {
  const candidate = (value ?? "").trim();
  if (candidate.startsWith("/") && !candidate.startsWith("//")) return candidate;
  return "/recommend";
}

function stateTtlSeconds(): number {
  const raw = Number(envValue("GOOGLE_LOGIN_STATE_TTL_SECONDS"));
  return Number.isSafeInteger(raw) && raw >= 60 ? raw : 600;
}

export function stateTtl(): number {
  return stateTtlSeconds();
}

function pkceChallenge(verifier: string): string {
  return createHash("sha256").update(verifier, "ascii").digest("base64url");
}

/** Build the Google authorization URL and the pending flow bound to the browser. */
export function startAuthorization(nextPath: string | null | undefined): { authorizationUrl: string; flow: PendingLoginFlow } {
  const client = loadLoginClient();
  const state = randomBytes(24).toString("base64url");
  const verifier = randomBytes(48).toString("base64url");
  const params = new URLSearchParams({
    client_id: client.clientId,
    redirect_uri: client.redirectUri,
    response_type: "code",
    scope: SCOPES,
    state,
    code_challenge: pkceChallenge(verifier),
    code_challenge_method: "S256",
    prompt: "select_account",
  });
  return {
    authorizationUrl: `${client.authUri}?${params.toString()}`,
    flow: { state, verifier, nextPath: safeNextPath(nextPath) },
  };
}

/**
 * Exchange the one-time authorization code and verify the returned ID token.
 * Returns the verified identity; the caller persists the local member.
 */
// fallow-ignore-next-line complexity -- Exchange errors and missing ID tokens are distinct failure modes of one boundary.
export async function completeAuthorization(code: string, verifier: string): Promise<GoogleIdentity> {
  const client = loadLoginClient();
  let payload: Record<string, unknown>;
  try {
    payload = await postForm(
      client.tokenUri,
      {
        code,
        client_id: client.clientId,
        client_secret: client.clientSecret,
        redirect_uri: client.redirectUri,
        grant_type: "authorization_code",
        code_verifier: verifier,
      },
      "Google token exchange failed",
    );
  } catch (error) {
    throw new GoogleLoginOAuthError(error instanceof Error ? error.message : "Google token exchange failed");
  }
  const idToken = String(payload.id_token ?? "").trim();
  if (!idToken) throw new GoogleLoginOAuthError("Google did not return an identity token");
  return verifyIdToken(idToken, client);
}

// fallow-ignore-next-line complexity -- Signature, audience, issuer, and claim checks each reject a distinct forgery.
async function verifyIdToken(idToken: string, client: GoogleLoginClient): Promise<GoogleIdentity> {
  let claims;
  try {
    const jwks = createRemoteJWKSet(new URL(client.jwksUri));
    const verified = await jwtVerify(idToken, jwks, {
      audience: client.clientId,
      issuer: [...GOOGLE_ISSUERS],
    });
    claims = verified.payload as Record<string, unknown>;
  } catch {
    throw new GoogleLoginOAuthError("Google identity token could not be verified");
  }
  if (claims.email_verified !== true) {
    throw new GoogleLoginOAuthError("Google account email is not verified");
  }
  const subject = String(claims.sub ?? "").trim();
  const email = String(claims.email ?? "").trim().toLowerCase();
  if (!subject || !email) throw new GoogleLoginOAuthError("Google identity is missing required fields");
  return {
    subjectId: subject,
    email,
    displayName: String(claims.name ?? "").trim(),
    avatarUrl: String(claims.picture ?? "").trim(),
  };
}
